core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java

```diff
@@ -327,27 +339,94 @@ private String getConfigurationFileName() {
         ClassLoader classLoader = TelemetryConfigurationFactory.class.getClassLoader();
         URL resource = classLoader.getResource(fileToParse);
 
-        // If not found as a resource, trying to load from the executing jar directory
         if (resource == null) {
-            try {
-                String jarFullPath = TelemetryConfigurationFactory.class.getProtectionDomain().getCodeSource().getLocation().toURI().getPath();
-                File jarFile = new File(jarFullPath);
+            InternalLogger.INSTANCE.logConfig(InternalLogger.LoggingLevel.INFO, "Did not find resource configuration file");
 
-                if (jarFile.exists()) {
-                    String jarDirectory = jarFile.getParent();
-                    String configurationPath = jarDirectory + File.separator + fileToParse;
-
-                    InternalLogger.INSTANCE.trace("Found configuration file: '%s'", configurationPath);
-                    return configurationPath;
-                }
-            } catch (URISyntaxException e) {
-                InternalLogger.INSTANCE.error("Failed to find configuration file: '%s'", e.getMessage());
+            // If not found as a resource, trying to load from the executing jar directory
+            String configurationFile = getConfFromSourceLocation();
+            if (configurationFile != null) {
+                return configurationFile;
             }
+
+            // If not found as a resource, trying to load from the class path
+            return getConfFromClassPath(classLoader);
         } else {
             String configurationFile = resource.getFile();
-            InternalLogger.INSTANCE.trace("Found configuration file: '%s'", configurationFile);
+            InternalLogger.INSTANCE.logConfig(InternalLogger.LoggingLevel.INFO, "Found configuration file: '%s'", configurationFile);
             return configurationFile;
         }
+    }
+
+    private String getConfFromClassPath(ClassLoader classLoader) {
+        if (!(classLoader instanceof URLClassLoader)) {
+            return null;
+        }
+
+        HashSet<String> checkedUrls = new HashSet<String>();
+
+        URL[] urls = ((URLClassLoader)classLoader).getURLs();
+        for (URL url : urls) {
+            String configurationPath = url.getPath();
+            if (configurationPath.endsWith(".jar")) {
+                int index = configurationPath.lastIndexOf('/');
+                if (index != -1) {
+                    configurationPath = configurationPath.substring(0, index + 1);
+                } else {
+                    continue;
+                }
+            } else {
+                if (!configurationPath.endsWith("/")) {
+                    configurationPath += '/';
+                }
+            }
+
+            if (checkedUrls.contains(configurationPath)) {
+                continue;
+            }
+
+            String configurationFile = configurationFileExists(configurationPath);
```

Comment:
I would expect a method named `configurationFileExists` to return a Boolean and not a string. Please rename this function so it would better reflect what it does and returns.